maint_tools/check_gha_workflow.py

```diff
@@ -0,0 +1,263 @@
+
+# can be found out at
+# "https://api.github.com/repos/{USER}/{REPO}/actions/workflows"
+WORKFLOW_ID = "71549417"
+
+INCLUDE_FAILED_RUNS = True
+
+# Pages of runs to collect
+# 100 per page
+PAGES_TO_COLLECT = range(1, 20)
+
+# If False, just plots the content of the TSV
+UPDATE_TSV = True
+
+OUTPUT_FILE = Path(__file__).parent / "test_runs_timing.tsv"
+
+
+def main(args=sys.argv) -> None:
+    """Collect duration of each job and plots them."""
+    update_tsv = UPDATE_TSV if OUTPUT_FILE.exists() else True
+
+    if update_tsv:
+
+        if len(args) > 1:
+            TOKEN = args[1]
+            auth = {"Authorization": "token " + TOKEN}
+        else:
+            auth = get_auth(USERNAME, TOKEN_FILE)
+
+        jobs_data = {"name": [], "started_at": [], "completed_at": []}
+
+        for page in PAGES_TO_COLLECT:
+            runs = get_runs(
+                WORKFLOW_ID,
+                auth,
+                page=page,
+                include_failed_runs=INCLUDE_FAILED_RUNS,
+            )
+            if len(runs) > 0:
+                print(f" found {len(runs)} runs")
+                jobs_data = udpate_jobs_data(jobs_data, runs, auth)
+            else:
+                break
+
+        df = pd.DataFrame(jobs_data)
+        df.to_csv(OUTPUT_FILE, sep="\t", index=False)
+
+    df = pd.read_csv(
+        OUTPUT_FILE,
+        sep="\t",
+        parse_dates=["started_at", "completed_at"],
+    )
+
+    df["duration"] = (df["completed_at"] - df["started_at"]) / pd.Timedelta(
+        minutes=1
+    )
+    df["python"] = df["name"].apply(set_python_version)
+    df["OS"] = df["name"].apply(set_os)
+    df["dependencies"] = df["name"].apply(set_dependencies)
+
+    print(df)
+
+    plot_job_durations(df)
+
+
+def plot_job_durations(df: pd.DataFrame) -> None:
+    """Plot and save."""
+    fig = px.line(
+        df,
+        x="started_at",
+        y="duration",
+        color="python",
+        symbol="dependencies",
+        line_dash="OS",
+        labels={
+            "duration": "Run duration (minutes)",
+            "started_at": "Run started on",
+            "OS": "OS",
+            "python": "python version",
+        },
+        title="Duration of nilearn test runs",
+    )
+
+    fig.update_xaxes(dtick="M1", tickformat="%b\n%Y")
+    fig.update_layout(autosize=True, width=1000, height=700)
+
+    fig.write_image(OUTPUT_FILE.with_suffix(".png"))
+    fig.write_html(OUTPUT_FILE.with_suffix(".html"))
+
+
+def set_os(x: str) -> str:
+    """Detect which OS the job was run on."""
+    if "ubuntu" in x:
+        return "ubuntu"
+    elif "windows" in x:
+        return "windows"
+    elif "macos" in x:
+        return "macos"
+    else:
+        return "n/a"
+
+
+def set_python_version(x: str) -> str:
+    """Detect which python version the job was run on."""
+    return next(
+        (
+            version
+            for version in ["3.8", "3.9", "3.10", "3.11", "3.12"]
```

Comment:
Is there a way we could keep this in-sync with the Python versions we support, without having to remember to update it here ?

Reply:
Cannot think of an easy way... Yet.

Reply:
Could not find an easy way to do this, so for now I at least extract this list into a constant at the top of the script that should make it easier to find and modify.